packages/custom_lint/lib/src/output.dart

```diff
@@ -0,0 +1,269 @@
+import 'package:cli_util/cli_logging.dart';
+import 'package:collection/collection.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as p;
+
+/// The maximum length of any of the existing severity labels.
+const int _severityWidth = 7;
+
+/// The number of spaces needed to indent follow-on lines (the body) under the
+/// message. The width left for the severity label plus the separator width.
+const int _bodyIndentWidth = _severityWidth + 3;
+
+final String _bodyIndent = ' ' * _bodyIndentWidth;
+
+/// is used as the usageLineLength.
+int? get _dartdevUsageLineLength =>
+    stdout.hasTerminal ? stdout.terminalColumns : null;
+
+/// Renders a list of lints in the given format.
+@internal
+void renderLints(
+  Logger log,
+  List<AnalysisErrorsParams> lints, {
+  required Directory workingDirectory,
+  required String format,
+}) {
+  final errors = lints
+      .expand(
+        (lintsForFile) => lintsForFile.errors
+          ..sort((a, b) {
+            final lineCompare =
+                a.location.startLine.compareTo(b.location.startLine);
+            if (lineCompare != 0) return lineCompare;
+            final columnCompare =
+                a.location.startColumn.compareTo(b.location.startColumn);
+            if (columnCompare != 0) return columnCompare;
+
+            final codeCompare = a.code.compareTo(b.code);
+            if (codeCompare != 0) return codeCompare;
+
+            return a.message.compareTo(b.message);
+          }),
+      )
+      .sortedBy((e) => _relativeFilePath(e.location.file, workingDirectory))
+      .sortedBy<num>((e) => -AnalysisErrorSeverity.VALUES.indexOf(e.severity));
+
+  if (errors.isEmpty) {
+    log.writeln('No issues found!');
+    return;
+  }
+
+  switch (format) {
+    case 'json':
+      emitJsonFormat(log, errors);
+      break;
+    default:
+      emitDefaultFormat(log, errors, workingDirectory);
+      break;
+  }
+}
+
+/// Emits a list of lints in the Dart analyzer style JSON format.
+@internal
+void emitJsonFormat(
+  Logger log,
+  Iterable<AnalysisError> errors,
+) {
+  Map<String, dynamic> location(String filePath, Map<String, dynamic> range) =>
+      {
+        'file': filePath,
+        'range': range,
+      };
+
+  Map<String, dynamic> position(int? offset, int? line, int? column) => {
+        'offset': offset,
+        'line': line,
+        'column': column,
+      };
+
+  Map<String, dynamic> range(
+    Map<String, dynamic> start,
+    Map<String, dynamic> end,
+  ) =>
+      {
+        'start': start,
+        'end': end,
+      };
+
+  final diagnostics = <Map<String, dynamic>>[];
+  for (final error in errors) {
+    final contextMessages = <Map<String, dynamic>>[];
+    if (error.contextMessages != null) {
+      for (final contextMessage in error.contextMessages!) {
+        final startOffset = contextMessage.location.offset;
+        contextMessages.add({
+          'location': location(
```

Comment:
All of this serialization logic sounds redundant with the AnalyzerConverter class. Which allows converting an AnalysisError from analyzer into one from analyzer_plugin. And the latter has a toJson

Reply:
Same as above, this is just taken from Dart analyzer and is different from the toJson output I believe.